go/logic/applier.go

```diff
@@ -76,12 +77,21 @@ func (this *Applier) InitDBConnections() (err error) {
 
 // validateConnection issues a simple can-connect to MySQL
 func (this *Applier) validateConnection(db *gosql.DB) error {
-	query := `select @@global.port, @@global.version`
+	extraPortRegexp := regexp.MustCompile("Percona|Maria")
+	query := `select @@global.port, @@global.version, @@global.version_comment`
 	var port int
-	if err := db.QueryRow(query).Scan(&port, &this.migrationContext.ApplierMySQLVersion); err != nil {
+	var extraPort int
+	var versionComment string
+	if err := db.QueryRow(query).Scan(&port, &this.migrationContext.ApplierMySQLVersion, &versionComment); err != nil {
 		return err
 	}
-	if port != this.connectionConfig.Key.Port {
+	if extraPortRegexp.MatchString(versionComment) {
+		extraPortQuery := `select @@global.extra_port`
+		if err := db.QueryRow(extraPortQuery).Scan(&extraPort); err != nil {
+			return err
+		}
+	}
```

Comment:
Are all Percona/MariaDB servers guaranteed to have this variable? If so, since which version?

Reply:
I'll do some digging. I believe Percona ported Maria's code and Maria has had it a while, but I'll try and get some concrete dates/version numbers.

Regardless, I do like your thoughts on brute forcing it. Then we don't even have to check for Percona or Maria (or the early versions without the functionality).

I'll update once I get more info.

Reply:
Looks like this was introduced in MariaDB  5.5.21 on 16 Mar 2012:

* https://mariadb.com/kb/en/library/mariadb-5521-release-notes/ (this was the beta release for the 5.5 line, but the functionality stuck around for all 5.5+ versions)
* https://mariadb.com/kb/en/library/changes-improvements-in-mariadb-55/
* https://mariadb.com/kb/en/library/thread-pool-in-mariadb/

I found this bit interesting:

> Significantly more efficient thread pool, comparable in functionality to the closed source feature in MySQL Enterprise.

So I assume it MySQL enterprise has the same extra_port semantics. I don't have access to it, so I don't know. I also don't see any extra port information in their public documentation https://dev.mysql.com/doc/refman/5.5/en/thread-pool.html

---

For Percona it was introduced at the same time with version 5.5.29-30.0 on 26 Feb 2012 as beta functionality:

* https://www.percona.com/doc/percona-server/5.5/release-notes/Percona-Server-5.5.29-30.0.html#5.5.29-30.0
* Percona: https://www.percona.com/doc/percona-server/5.5/performance/threadpool.html

But has persisted in 5.6 and 5.7 as a stable feature.

Percona explicitly mentions that it was ported from Maria's code:

> Ported the Thread Pool patch from MariaDB. This feature enables the server to keep the top performance even with the increased number of client connections.

I found the timeline interesting (Percona's release/announcement before Maria). But I assume some folks on Percona's side were keeping up to date on the Maria alpha/pre-merged improvements and brought it over to Percona early.

---

Completely off topic, but has GitHub evaluated using the Thread Pool for their own workload? I'm just assuming you all don't since you haven't run into this problem. I also imagine you all deal with MANY more connections than we do (total connections that is. you may have better partitioning so per-server connections are around the same).

I'm genuinely curious if you all evaluated it and decided against it for one reason or another. Your experience might influence us to switch back to one-connection-per-thread. You also might be running MySQL Community so this isn't even an option.

If you prefer not to answer that publicly, but still want to share your reasoning, by all means you can also shoot me an email at zach.moazeni@gmail.com

I have been meaning to write up a blog post but we noticed issues in our upgrade from Percona 5.6 to Percona 5.7. I worked with Percona's team but it's really hard to narrow down what exactly changed behind the scenes in 5.7. My assumption is that something that was blocking concurrency internally in 5.6 was unblocked in 5.7 and the more threads began running concurrently causing us load issues. Particularly on deploys when we cycle thousands of application connections (close down old ones for dying unicorns and spin up new ones for new unicorns).

We could have spent more time on the connection cycling (more sleep/pause between a unicorn dying and a unicorn taking its place) but moving to the thread pool was a decent solution and so far it has worked well for us. Rails 5's schema caching has also helped prevent a thundering herd of queries as our Rails instances boot up, which helped reduce the query load from connections as they startup.

Reply:
> Completely off topic, but has GitHub evaluated using the Thread Pool for their own workload? I'm just assuming you all don't since you haven't run into this problem. I also imagine you all deal with MANY more connections than we do (total connections that is. you may have better partitioning so per-server connections are around the same).

> I'm genuinely curious if you all evaluated it and decided against it for one reason or another. Your experience might influence us to switch back to one-connection-per-thread. You also might be running MySQL Community so this isn't even an option.

Nothing to hide. No, we haven't, and the reason is just backlog. There's a lot of MySQL features we just haven't had time to poke _yet_. Sometime...